File: android/Detonator/src/main/java/com/iconshot/detonator/renderer/Edge.java
package com.iconshot.detonator.renderer;

import com.iconshot.detonator.element.Element;

import java.util.List;

public class Edge {
    public int id;
    public Integer parent;
    public String contentType;
    public String attributes;
    public List<Edge> children;
    public String text;
    public boolean skipped;
    public boolean moved;

    public Element element;

    public int targetViewsCount;

    public Edge clone() {
        Edge edge = new Edge();

        edge.copyFrom(this);

        return edge;
    }

    public void copyFrom(Edge edge) {
        id = edge.id;
        parent = edge.parent;
        contentType = edge.contentType;
        attributes = edge.attributes;
        children = edge.children;
        text = edge.text;
        skipped = edge.skipped;
        moved = edge.moved;
        element = edge.element;
        targetViewsCount = edge.targetViewsCount;
    }
}

File: ios/Detonator/Sources/Detonator/Renderer/Edge.swift
import UIKit

public class Edge: Decodable {
    public var id: Int
    public var parent: Int?
    public var contentType: String?
    public var attributes: String?
    public var children: [Edge]
    public var text: String?
    public var skipped: Bool
    public var moved: Bool
    
    public var element: Element?
    
    public var targetViewsCount: Int = 0
    
    init(
        id: Int,
        parent: Int?,
        contentType: String?,
        attributes: String?,
        children: [Edge],
        text: String?,
        skipped: Bool,
        moved: Bool,
        element: Element?,
        targetViewsCount: Int
    ) {
        self.id = id
        self.parent = parent
        self.contentType = contentType
        self.attributes = attributes
        self.children = children
        self.text = text
        self.skipped = skipped
        self.moved = moved
        self.element = element
        self.targetViewsCount = targetViewsCount
    }
    
    public func clone() -> Edge {
        return Edge(
            id: id,
            parent: parent,
            contentType: contentType,
            attributes: attributes,
            children: children,
            text: text,
            skipped: skipped,
            moved: moved,
            element: element,
            targetViewsCount: targetViewsCount
        )
    }
    
    public func copyFrom(edge: Edge) {
        id = edge.id
        parent = edge.parent
        contentType = edge.contentType
        attributes = edge.attributes
        children = edge.children
        text = edge.text
        skipped = edge.skipped
        moved = edge.moved
        element = edge.element
        targetViewsCount = edge.targetViewsCount
    }
    
    private enum CodingKeys: String, CodingKey {
        case id
        case parent
        case contentType
        case attributes
        case children
        case text
        case skipped
        case moved
    }
}
